index.html

```diff
@@ -3,6 +3,14 @@
 <head>
     <meta charset="utf-8" />
     <script src="https://cdn.jsdelivr.net/gh/AssemblyScript/wabt.js@1.0.20/index.js"></script>
+    <link rel="stylesheet" href="../theme/css/test.css">
+    <link rel="stylesheet" href="../node_modules/codemirror/lib/codemirror.css">
+    <link rel="stylesheet" href="../node_modules/codemirror/theme/neo.css">
+    <script src="webstart.js"></script>
+    <script src="../node_modules/codemirror/lib/codemirror.js"></script>
+    <script src="../node_modules/codemirror/addon/edit/closebrackets.js"></script>
+    <script src="../node_modules/codemirror/mode/python/python.js"></script>
+    <script src="../theme/js/test.js"></script>
```

Comment:
I'm somewhat suspicious of including codemirror by directly referencing from node_modules. Was it not possible to include through webstart and making sure css is being loaded through webpack? (Codemirror docs on using a bundler. While for Rollup, some of this should still be reasonably transferrable https://codemirror.net/6/examples/bundle/)

Reply:
You are right that this is not the best practice of exposing server directory structure. For the sake of using stylesheets and JS I did not expect this to be an issue but you are right this is less than ideal. I can look into the docs you linked but FYI bundling could affect our future implementations for other UI features. Thanks for the info!

Reply:
Yeah this will actually break the deploy as-is, I think, because we don't ship a full source repo, just the `build/` directory, to the deploy repo. I'm not bothered by direct links like this, but may need to at least copy the files into `build/` as part of `npm run build-web`.

Reply:
@shanliyu Sounds good. I just want to make sure there is a good, clear reason for including an additional build system. My understanding so far is that gulp is ~only being used to process the scss transformations and provide hot reloading, both of which webpack already provides with the sass-loader plugin https://webpack.js.org/loaders/sass-loader/ and HMR https://webpack.js.org/concepts/hot-module-replacement/.

On the point made by @jpolitz, it's not at all clear how to run the build system for these front end changes since they evidently are not being invoked using webpack and `build-web`.

Reply:
If we are going to load codemirror sources directly, we should probably do it using a cdn (https://www.jsdelivr.com/package/npm/codemirror) and declaring it as an external in `webpack.config.js` in much the same way we already do for `wabt`
https://github.com/ucsd-cse231-w21/chocopy-wasm-compiler/blob/3ae392fe4f1d13f2b2432e6fab2ddcc96131d48d/index.html#L5-L8
https://github.com/ucsd-cse231-w21/chocopy-wasm-compiler/blob/3ae392fe4f1d13f2b2432e6fab2ddcc96131d48d/webpack.config.js#L18-L20

Reply:
@R167 To answer your question, I personally think integrating sass transformation into build slows down front-end development significantly, especially when more and more changes are being added to the compiler. But using gulp it is independent from the build process which is much faster for us during development

Reply:
@jpolitz Can you elaborate a bit more on the deployment process, such as the server structure? I was assuming the hosting server contains everything including `node_modules` as CI/CD builds the site so I included the stylesheets and script tags this way

Reply:
> I personally think integrating sass transformation into build slows down front-end development significantly

I agree that build slows down dev. That's we we're using webpack-dev-server now. https://github.com/ucsd-cse231-w21/chocopy-wasm-compiler/pull/1 Rebuilds take <0.5s. (launch using `npm start` as mentioned in the README).

> I was assuming the hosting server contains everything including `node_modules` as CI/CD builds the site so I included the stylesheets and script tags this way

No, `npm run build-web` only runs `webpack` in production mode. The gh-pages deploy phase then pushes the contents of the `./build` directory to the `gh-pages` branch which is then severed by GitHub pages.

Reply:
example branch which moves code mirror and sass to being loaded by webpack https://github.com/ucsd-cse231-w21/chocopy-wasm-compiler/compare/main...R167:frontend-example

Reply:
example branch which moves code mirror and sass to being loaded by webpack https://github.com/ucsd-cse231-w21/chocopy-wasm-compiler/compare/main...R167:frontend-example

Reply:
Thanks for the help but this branch does not seem to build
```ERROR in ./webstart.ts 6:0-22
Module not found: Error: Can't resolve './style.scss' in '/Users/shanliyu/School/CSE231/chocopy-wasm-compiler'
resolve './style.scss' in '/Users/shanliyu/School/CSE231/chocopy-wasm-compiler'
  using description file: /Users/shanliyu/School/CSE231/chocopy-wasm-compiler/package.json (relative path: .)
    Field 'browser' doesn't contain a valid alias configuration
    using description file: /Users/shanliyu/School/CSE231/chocopy-wasm-compiler/package.json (relative path: ./style.scss)
      no extension
        Field 'browser' doesn't contain a valid alias configuration
        /Users/shanliyu/School/CSE231/chocopy-wasm-compiler/style.scss doesn't exist
      .ts
        Field 'browser' doesn't contain a valid alias configuration
        /Users/shanliyu/School/CSE231/chocopy-wasm-compiler/style.scss.ts doesn't exist
      .js
        Field 'browser' doesn't contain a valid alias configuration
        /Users/shanliyu/School/CSE231/chocopy-wasm-compiler/style.scss.js doesn't exist
      as directory
        /Users/shanliyu/School/CSE231/chocopy-wasm-compiler/style.scss doesn't exist```
I am exploring the option you provided by using CDN rather than locally for codemirror, and I will still attempt to use gulp to concatenate and minify CSS/JS which will be copied into `/build`.

Reply:
@shanliyu What checkout/build commands did you run to get that to happen? Running from that branch worked fine on my end(?)
```
git remote add R167 git@github.com:R167/chocopy-wasm-compiler.git
git fetch R167
git checkout R167/frontend-example
npm start
```

Reply:
So instead of checking to your branch, I copied all of your files and modified my current repo (6 files changed from the link), but when running `npm run build-web` the above error occurs. I tried switching to your branch and it worked without any issues so a bit confused

Reply:
best guess is you maybe missed a line or didn't restart webpack after npm install? if you want to use any of this (merging probably easiest), feel free.

Reply:
Really appreciate the help Winston, after merging with the branch it seems that the error is resolved. Not sure why it was occurring before. Thanks again!